### DISCUSSION TITLE
[QUESTION] How to run longhorn engine v2 without containers

### DISCUSSION THREAD

**Discussion by hookak** [longhorn/longhorn#7818] [Q&A]
## Question

Hello, I am very interested in longhorn engine v2.
Recently, I saw that you are working on integrating the engine v2 code into the spdk,
Is there any documentation to use engine v2 from spdk source build without using container?
(https://github.com/longhorn/longhorn/wiki/How-to-run-longhorn-engine-without-containers for v1)
The existing ones in longhorn/spdk, lognhorn/longhorn-spdk-engine seem to be outdated.

Thanks in advance

**Comment by c3y1huang** [comment]
cc @derekbit , @shuo-wu

**Comment by derekbit** [comment]
It might be not easy to run an engine without a container, because `spdk_tgt` daemon is specifically designed to operate within an instance-manager pod.
Could you elaborate more on the purpose of the request and application scenario?

**Comment by hookak** [comment]
My purpose is to build and test longhorn engine v2(to see the performance difference between vanila spdk + nvmeOF).
If it is not possible to operate without an `instance-manager pod`,
could we know how you build and test engine v2 source code?(especifically, the local replica mode in engine v2) @derekbit

**Comment by shuo-wu** [comment]
Running longhorn engine v2 without containers means you need to manually:
1. install [the dependencies](https://github.com/longhorn/longhorn-instance-manager/blob/master/package/Dockerfile)
2. build and run `spdk_tgt` daemon
3. start the Longhorn gRPC SPDK server
4. add disks to the Longhorn SPDK server via [the gRPC call](https://github.com/longhorn/longhorn-spdk-engine/blob/c5f14845bd830f948cc98b7bfc88a56cddd68511/proto/spdkrpc/spdk.proto#L49)
5. launch engines and replicas via [the gRPC calls](https://github.com/longhorn/longhorn-spdk-engine/blob/c5f14845bd830f948cc98b7bfc88a56cddd68511/proto/spdkrpc/spdk.proto#L9). For example: 
    ```
    grpcurl -d '' -plaintext -import-path longhorn-spdk-engine/proto/spdkrpc/ -proto longhorn-spdk-engine/proto/spdkrpc/spdk.proto -d '{"name":"test-r-1", "lvs_name":"spdk-disk", "lvs_uuid":"bf13bb1f-d197-4eb8-80ff-60eabfbe9a41", "spec_size":2147483648, "port_count":5, "expose_required":false}' 172.17.0.2:8504 spdkrpc.SPDKService/ReplicaCreate
    grpcurl -d '' -plaintext -import-path ./proto/spdkrpc/ -proto ./proto/spdkrpc/spdk.proto -d '{"name":"test-e", "volume_name":"test", "spec_size":2147483648, "replica_address_map":{"test-r-1":"172.17.0.2:20001"}, "frontend": "spdk-tcp-blockdev", "port_count": 1}' 172.17.0.2:8504 spdkrpc.SPDKService/EngineCreate
    ```

To be honest, it's laborious to figure out all of this from zero and tune the gRPC call parameters.

**Reply by hookak** [reply]
Could you give me a little more detail on step 3?
Running `make` on the `longhorn-spdk-engine` source code makes a `longhorn-spdk-engine:main` docker image.
Could you tell me how to run `Longhorn gRPC SPDK server` through that docker image? @shuo-wu

**Reply by shuo-wu** [reply]
IIRC there is no cmdline to launch the Longhorn gRPC SPDK server. You can check [this](https://github.com/longhorn/longhorn-spdk-engine/blob/c5f14845bd830f948cc98b7bfc88a56cddd68511/pkg/spdk_test.go#L94) as reference.

**Comment by hookak** [comment]
Currently I am not using longhorn-spdk-engine, I am using jsonrpc (rpc.py) from longhorn/spdk to create devices with the same architecture as longhorn engine v2.

I am following Damiano's presentation at SUSECON 2023 (https://www.youtube.com/watch?v=oTs_2AF6JN8) to create the same.

Here is what I have done
```
./rpc.py bdev_aio_create /dev/nvme0n1 disk0 
./rpc.py bdev_lvol_create_lvstore disk0 lvs0
./rpc.py bdev_lvol_create replica0 10240 -l lvs0
./rpc.py bdev_raid_create -r raid1 -b bdev_replica0,bdev_replica1 -n volume0
./rpc.py bdev_raid_get_bdevs all
[
  {
    "name": "volume0",
    "uuid": "00000000-0000-0000-0000-000000000000",
    "strip_size_kb": 0,
    "state": "configuring",
    "raid_level": "raid1",
    "superblock": false,
    "num_base_bdevs": 1,
    "num_base_bdevs_discovered": 0,
    "num_base_bdevs_operational": 1,
    "base_bdevs_list": [
      {
        "name": "bdev_replica0,bdev_replica1",
        "uuid": "00000000-0000-0000-0000-000000000000",
        "is_configured": false,
        "data_offset": 0,
        "data_size": 0
      }
    ]
  }
]
```

Where I am stuck is represented by Question1,2 in the Architecture in the figure.
<img width="1027" alt="image" src="https://github.com/longhorn/longhorn/assets/21308664/2ffed3f7-5e41-4399-8855-731647edf06d">


Question 1. how do I create a bdev(bdev_replica0) with lvol(replica0) as backend?
Question 2. How do I create a bdev(bdev_replica1) with lvol(replica1) as backend on remote Node?

Can I solve 2 questions using rpc.py in longhorn/spdk? I need some help!. @DamiaSan 


Thanks in advance

**Reply by shuo-wu** [reply]
You can check this doc but I am not sure if it still works: https://github.com/longhorn/longhorn-spdk-engine/wiki/How-to-setup-a-RAID1-block-device-with-SPDK

**Reply by DamiaSan** [reply]
Question 1: simply pass `lvs0/lvol0`, instead `bdev_replica0` which does not exist, to the raid creation as base bdev
Question 2: you have to create another lvol on a remote node, export it via NVMe-oF, attach to it on the node where the raid resides and use this attached bdev as base bdev for the raid.

[Here](https://github.com/longhorn/longhorn-spdk-engine/wiki/How-to-setup-a-RAID1-block-device-with-SPDK) you can find a complete description of the step needed.

**Reply by hookak** [reply]
Now I understand it completely. I didn't check https://github.com/longhorn/longhorn-spdk-engine/wiki/How-to-setup-a-RAID1-block-device-with-SPDK.
I'll check it out and contribute if there are any gaps.
Really appreciated for your time. @shuo-wu @DamiaSan

**Comment by gaby** [comment]
On a similar topic, is it possible to run Longhorn without Kubernetes?

**Comment by hookak** [comment]
Hi @gaby, it's better to create a new question for that
> is it possible to run Longhorn without Kubernetes?

Close this discussion
